File: examples/example_usage.cpp
#include "databento/dbn_parser.hpp"
#include <iostream>
#include <chrono>

using namespace databento;

void example_direct_api(const std::string& filename) {
    std::cout << "=== Direct API Example (330M records/sec) ===" << std::endl;
    
    auto start = std::chrono::high_resolution_clock::now();
    
    auto records = parse_file_mbo_fast(filename);
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::cout << "Parsed " << records.size() << " records in " 
              << duration.count() << " ms" << std::endl;
    
    if (!records.empty()) {
        std::cout << "First record - Order ID: " << records[0].order_id 
                  << ", Price: " << records[0].price 
                  << ", Size: " << records[0].size << std::endl;
    }
}

void example_batch_api(const std::string& filename) {
    std::cout << "\n=== Batch API Example (150M records/sec) ===" << std::endl;
    
    auto start = std::chrono::high_resolution_clock::now();
    
    auto records = parse_file_mbo_batch(filename, 10000);
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::cout << "Parsed " << records.size() << " records in " 
              << duration.count() << " ms" << std::endl;
}

void example_callback_api(const std::string& filename) {
    std::cout << "\n=== Callback API Example (30M records/sec) ===" << std::endl;
    
    size_t count = 0;
    int64_t total_size = 0;
    
    auto start = std::chrono::high_resolution_clock::now();
    
    parse_file_mbo_callback(filename, [&count, &total_size](const MBOMsg& msg) {
        count++;
        total_size += msg.size;
        
        // Process each record here
        if (count == 1) {
            std::cout << "First record - Order ID: " << msg.order_id << std::endl;
        }
    });
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::cout << "Processed " << count << " records in " 
              << duration.count() << " ms" << std::endl;
    std::cout << "Average size: " << (count > 0 ? total_size / count : 0) << std::endl;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <dbn_file>" << std::endl;
        return 1;
    }
    
    std::string filename = argv[1];
    
    try {
        example_direct_api(filename);
        example_batch_api(filename);
        example_callback_api(filename);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: examples/example_usage.py
#!/usr/bin/env python3
"""
Example usage of databento-fast Python bindings
"""

import databento_cpp
import time

def example_direct_api(filename):
    """Direct API - Fastest (330M records/sec)"""
    print("=== Direct API Example (330M records/sec) ===")
    
    start = time.time()
    records = databento_cpp.parse_file_mbo_fast(filename)
    end = time.time()
    
    print(f"Parsed {len(records)} records in {(end - start) * 1000:.2f} ms")
    
    if records:
        first = records[0]
        print(f"First record - Order ID: {first.order_id}, "
              f"Price: {first.price}, Size: {first.size}")
    
    return records

def example_batch_api(filename):
    """Batch API - Memory efficient (150M records/sec)"""
    print("\n=== Batch API Example (150M records/sec) ===")
    
    start = time.time()
    records = databento_cpp.parse_file_mbo_batch(filename, batch_size=10000)
    end = time.time()
    
    print(f"Parsed {len(records)} records in {(end - start) * 1000:.2f} ms")
    
    return records

def example_callback_api(filename):
    """Callback API - Event-driven (30M records/sec)"""
    print("\n=== Callback API Example (30M records/sec) ===")
    
    count = 0
    total_size = 0
    
    def process_record(record):
        nonlocal count, total_size
        count += 1
        total_size += record.size
        
        if count == 1:
            print(f"First record - Order ID: {record.order_id}")
    
    start = time.time()
    databento_cpp.parse_file_mbo_callback(filename, process_record)
    end = time.time()
    
    print(f"Processed {count} records in {(end - start) * 1000:.2f} ms")
    print(f"Average size: {total_size // count if count > 0 else 0}")

def main():
    import sys
    
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <dbn_file>")
        sys.exit(1)
    
    filename = sys.argv[1]
    
    try:
        example_direct_api(filename)
        example_batch_api(filename)
        example_callback_api(filename)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
